File: src/SpectraSpark/rmc/rmc2d.cpp
#include "rmc2d.hpp"
#include <iostream>
#include <omp.h>
#include <cmath>
#include <set>
#include <random>
#include <algorithm>
#include <cassert>

#define SQ(x) ((x)*(x))

namespace RMC {

const int Simulator2d::STEP_A[] = {1, 0, -1, -1, 0, 1};
const int Simulator2d::STEP_B[] = {0, 1, 1, 0, -1, -1};
const int Simulator2d::PROHIBITED_A_REL[] = {
    0, 1, 1,  0,  -1, -1, 0,                     // 1NN
    2, 1, -1, -2, -1, 1,                         // 2NN
    2, 2, 0,  -2, -2, 0,                         // 3NN
    3, 3, 2,  1,  -1, -2, -3, -3, -2, -1, 1, 2,  // 4NN
    3, 3, 0,  -3, -3, 0                          // 5NN
};
const int Simulator2d::PROHIBITED_B_REL[] = {
    0, 0, 1, 1,  0,  -1, -1,                      // 1NN
    1, 2, 1, -1, -2, -1,                          // 2NN
    0, 2, 2, 0,  -2, -2,                          // 3NN
    1, 2, 3, 3,  2,  1,  -1, -2, -3, -3, -2, -1,  // 4NN
    0, 3, 3, 0,  -3, -3                           // 5NN
};
int Simulator2d::max_rand_trial = 10000;

Simulator2d::Simulator2d(const int _seed) : seed(_seed), engine(_seed), dist(0, 1) {
  this->set_rotation(0);
}

Simulator2d::~Simulator2d() {
  delete[] a;
  delete[] b;
  delete[] prohibited_a;
  delete[] prohibited_b;
  delete[] x;
  delete[] y;
  delete[] a_re;
  delete[] a_im;
  delete[] qx;
  delete[] qy;
  delete[] i_exp;
  delete[] i_sim;
  delete[] i_par;
}

void Simulator2d::set_model(int _La, int _Lb, int _n, const int *_a, const int *_b) {
  this->La = _La;
  this->Lb = _Lb;
  this->n = _n;
  this->a = new int[n];
  this->b = new int[n];
  #pragma omp parallel for
  for (int i = 0; i < n; i++) {
    this->a[i] = _a[i];
    this->b[i] = _b[i];
  }

  this->prohibited_a = new int[n*N_PROHIBITED];
  this->prohibited_b = new int[n*N_PROHIBITED];
  #pragma omp parallel for
  for (int i = 0; i < n; i++) {
    this->update_prohibited(i, this->a[i], this->b[i]);
  }

  this->x = new double[n];
  this->y = new double[n];
  this->ab2xy(n, this->a, this->b, this->x, this->y);
}

void Simulator2d::set_exp_data(
  int w, int h, double *_qx, const double *_qy, const double *_i_exp
) {
  this->width = w, this->height = h;
  this->qx_seq = new double[w], this->qy_seq = new double[h];
  for (int i=0; i<w; i++) this->qx_seq[i] = _qx[i];
  for (int i=0; i<h; i++) this->qy_seq[i] = _qy[i];
  ulong n_all = w * h;
  this->map = new ulong[n_all];
  for (ulong i = 0; i < n_all; i++) {
    this->map[i] = -1;
  }
  ulong cur = 0;
  for (int y = 0; y < h; y++) {
    for (int x = 0; x < w; x++) {
      int idx = y * w + x;
      if (std::isfinite(_i_exp[idx])) {
        double _q2 = _qx[x] * _qx[x] + _qy[y] * _qy[y];
        if (_q2 < 1e-10) continue;
        this->map[y*w + x] = cur;
        cur++;
      }
    }
  }
  this->n_px = cur;

  this->qx = new double[this->n_px];
  this->qy = new double[this->n_px];
  this->i_exp = new double[this->n_px];
  this->i_sim = new double[this->n_px];
  this->a_re = new double[this->n_px];
  this->a_im = new double[this->n_px];
  this->i_par = new double[this->n_px];
  this->i_sum = 0, this->isq_sum = 0;
  cur = 0;
  for (ulong i = 0; i < n_all; i++) {
    int y = i / w, x = i % w;
    if (this->map[i] == -1) continue;
    ulong xy = y * w + x;
    this->qx[cur] = _qx[x], this->qy[cur] = _qy[y];
    this->i_exp[cur] = _i_exp[xy];
    this->i_sum += _i_exp[xy];
    this->isq_sum += SQ(_i_exp[xy]);
    double _w = sqrt(_qx[x] * _qx[x] + _qy[y] * _qy[y]) * this->R_PARTICLE;
    this->i_par[cur] = SQ(3 * (sin(_w) - _w * cos(_w)) / (_w * _w * _w));
    cur++;
  }
}

void Simulator2d::anneal(int n_iter, int n_move) {
  int *idx = new int[n_move];
  int *a_before = new int[n_move];
  int *b_before = new int[n_move];
  double *_d = new double[n_move];
  int i;
  try {
    for (i=0; i<n_iter; i++) {
      this->move(n_move, idx, a_before, b_before, _d, _d, _d, _d);
    }
  } catch (const std::runtime_error e) {
    std::cerr << e.what() << " at step " << i << std::endl;
  }
  this->ab2xy(n, this->a, this->b, this->x, this->y);
  delete[] idx;
  delete[] a_before;
  delete[] b_before;
  delete[] _d;
}

void Simulator2d::set_rotation(double _theta) {
  this->theta = _theta;
  _theta = _theta * M_PI / 180;
  double eax = cos(_theta), eay = sin(_theta);
  _theta += 2 * M_PI / 3;
  double ebx = cos(_theta), eby = sin(_theta);
  this->mat_ab2xy[0][0] = eax * this->A_MG;
  this->mat_ab2xy[0][1] = eay * this->A_MG;
  this->mat_ab2xy[1][0] = ebx * this->A_MG;
  this->mat_ab2xy[1][1] = eby * this->A_MG;
}

void Simulator2d::compute_i() {
  double _i_sum = 0;
  #pragma omp parallel for reduction(+:_i_sum)
  for (ulong i = 0; i < this->n_px; i++) {
    a_re[i] = 0;
    a_im[i] = 0;
    for (int j = 0; j < this->n; j++) {
      double delta = qx[i] * x[j] + qy[i] * y[j];
      a_re[i] += cos(delta);
      a_im[i] += sin(delta);
    }
    i_sim[i] = i_par[i] * (SQ(a_re[i]) + SQ(a_im[i]));
    _i_sum += i_sim[i];
  }

  double c = this->i_sum / _i_sum;
  #pragma omp parallel for
  for (ulong i = 0; i < this->n_px; i++) {
    i_sim[i] *= c;
  }
}

double Simulator2d::compute_residual() const {
  double residual = 0;
  #pragma omp parallel for reduction(+:residual)
  for (ulong i = 0; i < this->n_px; i++) {
    residual += SQ(i_exp[i] - i_sim[i]);
  }
  return residual / isq_sum;
}

int Simulator2d::run (
  int n_move, int max_iter, double *res_hist, double sigma2, double thresh
) {
  this->compute_i();
  res_hist[0] = this->compute_residual();
  double inv_sigma = 1 / sigma2;

  int *idx = new int[n_move];
  int *a_before = new int[n_move];
  int *b_before = new int[n_move];
  double *x_before = new double[n_move];
  double *y_before = new double[n_move];
  double *x_after = new double[n_move];
  double *y_after = new double[n_move];

  int i;
  try {
    for (i=1; i<=max_iter; i++) {
      this->move(n_move, idx, a_before, b_before, x_before, y_before, x_after, y_after);
      this->update_i(n_move, x_before, y_before, x_after, y_after);
      res_hist[i] = this->compute_residual();
      assert (res_hist[i] >= 0);
      if (res_hist[i] < res_hist[i - 1]) {
        // accept
        if (res_hist[i] < thresh) break;
      } else if (dist(engine) <
                exp(-(res_hist[i] - res_hist[i - 1]) * inv_sigma)) {
        // acceptして収束判定
        if (res_hist[i] < thresh) break;
      } else {
        // rejectして元に戻す
        res_hist[i] = res_hist[i-1];
        for (int j=0; j<n_move; j++) {
          this->a[idx[j]] = a_before[j];
          this->b[idx[j]] = b_before[j];
          this->update_prohibited(idx[j], a_before[j], b_before[j]);
        }
      }
    }
  } catch (const std::runtime_error e) {
    std::cerr << e.what() << " at step " << i << std::endl;
  }
  delete[] idx;
  delete[] a_before;
  delete[] b_before;
  delete[] x_before;
  delete[] y_before;
  delete[] x_after;
  delete[] y_after;
  return i;
}

void Simulator2d::move(
  int n_move, int *idx, int *a_before, int *b_before,
  double *x_before, double *y_before, double *x_after, double *y_after
) {
  // Not parallelized to ensure repeatability of random number generation
  std::set<int> idx_set;
  int dir[] = {0, 1, 2, 3, 4, 5};
  ulong n_prohibited = this->n * N_PROHIBITED;
  int all_success = 1;
  for (int trial=0; trial < this->max_rand_trial; trial++) {
    // choose non-repeating indices
    idx_set.clear();
    while (idx_set.size() < (ulong)n_move) {
      int idx_i = this->engine() % this->n;
      if (idx_set.find(idx_i) == idx_set.end()) {
        idx[idx_set.size()] = idx_i;
        idx_set.insert(idx_i);
      }
    }

    // save current position
    for (int i=0; i<n_move; i++) {
      a_before[i] = this->a[idx[i]];
      b_before[i] = this->b[idx[i]];
    }

    for (int i=0; i<n_move; i++) {
      // check if the move is possible
      int particle_success = 0;
      int toa, tob;
      std::shuffle(dir, dir+6, this->engine);
      for (int j=0; j<6; j++) {
        int overlap = 0;
        toa = (this->a[idx[i]] + STEP_A[dir[j]] + La) % La;
        tob = (this->b[idx[i]] + STEP_B[dir[j]] + Lb) % Lb;
        #pragma omp parallel for reduction(+:overlap)
        for (ulong k=0; k < n_prohibited; k++) {
          overlap += (toa == prohibited_a[k] && tob == prohibited_b[k]);
        }
        // overlap==1 means o.k.
        if (overlap == 1) {
          particle_success = 1;
          break;
        } else {
          continue;
        }
      }
      if (particle_success) {
        this->a[idx[i]] = toa;
        this->b[idx[i]] = tob;
        this->update_prohibited(idx[i], toa, tob);
      } else {
        all_success = 0;
        break;
      }
    }

    if (all_success) {
      break; // success
    } else {
      // revert and try again
      for (int i=0; i<n_move; i++) {
        this->a[idx[i]] = a_before[i];
        this->b[idx[i]] = b_before[i];
        this->update_prohibited(idx[i], a_before[i], b_before[i]);
      }
      continue;
    }
  }

  if (!all_success) {
    throw std::runtime_error("Runtime Error : Failed to move particles");
  };

  int *a_after = new int[n_move];
  int *b_after = new int[n_move];
  for (int i=0; i<n_move; i++) {
    int idx_i = idx[i];
    x_before[i] = this->x[idx_i];
    y_before[i] = this->y[idx_i];
    a_after[i] = this->a[idx_i];
    b_after[i] = this->b[idx_i];
  }
  this->ab2xy(n_move, a_after, b_after, x_after, y_after);
  delete[] a_after;
  delete[] b_after;
  return;
}

void Simulator2d::update_i(
  int n, double *x_before, double *y_before, double *x_after, double *y_after
) {
  // ulong n_px = width * height;
  for (int i = 0; i < n; i++) {
    double xb = x_before[i], yb = y_before[i], xa = x_after[i], ya = y_after[i];
    #pragma omp parallel for
    for (ulong j = 0; j < n_px; j++) {
      double delta_b = qx[j] * xb + qy[j] * yb;
      double delta_a = qx[j] * xa + qy[j] * ya;
      this->a_re[j] += -cos(delta_b) + cos(delta_a);
      this->a_im[j] += -sin(delta_b) + sin(delta_a);
    }
  }

  double i_sum = 0;
  #pragma omp parallel for reduction(+:i_sum)
  for (ulong i = 0; i < n_px; i++) {
    this->i_sim[i] = i_par[i] * (SQ(this->a_re[i]) + SQ(this->a_im[i]));
    i_sum += this->i_sim[i];
  }

  double c = this->i_sum / i_sum;
  #pragma omp parallel for
  for (ulong i = 0; i < n_px; i++) {
    this->i_sim[i] *= c;
  }
}

}  // namespace RMC

File: src/SpectraSpark/rmc/rmc2d.py
#! /usr/bin/python3

import numpy as np
from typing import overload

SQRT3 = np.sqrt(3)


class RmcSim2d:
    # 格子定数(nm)
    A_MG = 0.321
    # 粒子の半径(nm)
    R_PARTICLE = 0.355
    # 乱数を使った試行回数の上限
    MAX_RAND_TRIAL = 100
    # 粒子の移動方向ごとの変位(格子座標)
    STEP = ((1, 0), (1, 1), (0, 1), (-1, 0), (-1, -1), (0, -1))
    # 粒子近傍の禁止域の相対位置
    PROHIBITED = np.array([(
        0, 1, 1,  0,  -1, -1, 0,    # 0,1NN
        2, 1, -1, -2, -1, 1,        # 2NN
        2, 2, 0,  -2, -2, 0,        # 3NN
        3, 3, 2,  1,  -1, -2, -3, -3, -2, -1, 1, 2, # 4NN
        3, 3, 0,  -3, -3, 0         # 5NN
    ), (
        0, 0, 1, 1,  0,  -1, -1,    # 0,1NN
        1, 2, 1, -1, -2, -1,        # 2NN
        0, 2, 2, 0,  -2, -2,        # 3NN
        1, 2, 3, 3,  2,  1,  -1, -2, -3, -3, -2, -1, # 4NN
        0, 3, 3, 0,  -3, -3         # 5NN
)])

    def __init__(self, seed: int = 0):
        # モデル空間のサイズ
        self.La, self.Lb = 0, 0
        # モデル中の粒子数
        self.N = 0
        # フィッティングするqの配列
        self.qx, qy = np.array([]), np.array([])
        # 実験データ
        self.i_exp = np.array([])


        # 粒子の座標(格子座標:a=x, b=-0.5x+rt(3)/2 y)
        self.pos_ab = np.array([[np.nan, np.nan]])
        # 格子ベクトルの定義
        self.set_rot(0)
        # 禁止位置のab座標の配列 : インデックスは[粒子の番号,相対位置]の順
        self.prohibited_a, self.prohibited_b = np.array([]), np.array([])
        # 散乱振幅の実部、虚部
        self.a_re, self.a_im = np.array([]), np.array([])
        # 散乱強度
        self.i_sim = np.array([])
        self.i_sum = np.nan
        self.i_dtype = np.int32

        # 乱数生成器
        self.rs = np.random.RandomState(seed)
        return

    def run(self, n_move: int, max_iter: int,
            sigma2:float=1e-1, thresh: float = 1e-20) -> np.ndarray:
        """フィッティングを実行する

        Parameters
        ----------
        n_move: int
            1ステップで動かす粒子の数
        max_iter : int
            試行回数の上限
        thresh : float, optional
            停止する残差の閾値, by default 1e-20

        Returns
        -------
        np.ndarray
            residualの履歴
        """
        self.compute_i()
        res_hist = np.empty(max_iter + 1)
        res_hist[0] = self.compute_residual()
        sigma2 = sigma2 * res_hist[0]
        ran = self.rs.rand(max_iter)
        for i in range(max_iter):
            before_ab = self.pos_ab.copy()
            before_xy, after_xy = self.__move(n_move)
            self.update_i(before_xy, after_xy)
            new_res = self.compute_residual()
            d_res = new_res - res_hist[i]
            if (d_res<0) or ((np.exp(-d_res / sigma2) > ran[i])):
                pass # accept
            else:
                # rejectとして元に戻す
                self.pos_ab = before_ab
                self.update_i(after_xy, before_xy)
                new_res = res_hist[i]
            res_hist[i + 1] = new_res

            if new_res < thresh:
                res_hist = res_hist[:i + 2]
                break

        self.pos_xy = np.dot(self.pos_ab, self.conv)
        return res_hist

    def set_exp_data(self, data: np.ndarray, qx: np.ndarray, qy: np.ndarray,
                     *, dtype=np.int32, v_min:float=2.0, v_max:float=2**20, q_thresh:float=1.0):
        """散乱強度の実験値をセットする
        散乱強度は全要素の和が1になるように規格化する

        Parameters
        ----------
        data : np.ndarray
            二次元散乱強度プロフファイル
        qx : np.ndarray
            dataの第2軸に対応するqの配列
        qy : np.ndarray
            dataの第1軸に対応するqの配列
        dtype : np.dtype, optional
            散乱強度の型, by default np.int32
        v_min : float, optional
            強度がv_min以下の部分をマスクする, by default 2.0
        v_max : float, optional
            強度がv_max以上の部分をマスクする, by default 2**20
        q_thresh : float, optional
            qがq_thresh以下の部分をマスクする, by default 1.0
        """
        xx, yy = np.meshgrid(qx, qy)
        q = np.sqrt(xx**2 + yy**2)

        # マスクで隠す画素が0
        self.mask = ((data > v_min) * (data < v_max) * (q > q_thresh)).astype(np.int8)
        self.i_exp = (data * self.mask).astype(dtype)
        self.i_sum = np.sum(self.i_exp)
        self.i_dtype = dtype
        self.qx, self.qy = qx, qy

        w = q * self.R_PARTICLE
        self.i_par = (3 * (np.sin(w) - w * np.cos(w)) * w**(-3))**2
        self.i_par[w == 0] = 0
        return

    def set_model(self, La: int, Lb: int, pos_ab: np.ndarray):
        """モデル空間のサイズと粒子数を設定する

        Parameters
        ----------
        La : int
        Lb : int
        pos_ab : np.ndarray
        """
        assert pos_ab.shape[1] == 2
        assert np.all((pos_ab[:0] >= 0) * (pos_ab[:0] < La))
        assert np.all((pos_ab[:1] >= 0) * (pos_ab[:1] < Lb))
        self.La, self.Lb = La, Lb
        self.N = pos_ab.shape[0]
        self.pos_ab = pos_ab.copy()
        self.pos_xy = np.dot(self.pos_ab, self.conv)

        # 禁止位置の生成
        # xx: 各列に相対位置が入った配列
        # yy: 各行に粒子位置が入った配列
        xx, yy = np.meshgrid(self.PROHIBITED[:,0], self.pos_ab[:,0])
        self.prohibited_a = (xx + yy) % self.La
        xx, yy = np.meshgrid(self.PROHIBITED[:,1], self.pos_ab[:,1])
        self.prohibited_b = (xx + yy) % self.Lb
        return

    def anneal(self, n_iter:int=1_000, n_move=1):
        for _ in range(n_iter):
            self.__move(n_move=n_move)
        self.pos_xy = np.dot(self.pos_ab, self.conv)
        return

    def set_rot(self, rot: float):
        """格子ベクトルを開店する

        Parameters
        ----------
        rot : float
            a軸とx軸のなす角[deg]
        """
        ea, eb = np.array([1, 0]), np.array([-0.5, SQRT3/2])
        rot = np.radians(rot)
        rot_mat = np.array([
            [np.cos(rot), -np.sin(rot)],
            [np.sin(rot), np.cos(rot)]
        ])
        self.ea = np.dot(rot_mat, ea) * self.A_MG
        self.eb = np.dot(rot_mat, eb) * self.A_MG
        self.conv = np.vstack((self.ea, self.eb))
        self.pos_xy = np.dot(self.pos_ab, self.conv)
        return

    def compute_i(self):
        """散乱強度を計算する"""
        nx, ny, N = self.qx.size, self.qy.size, self.N
        qx, qy = np.meshgrid(self.qx, self.qy)
        # Kahanの加算アルゴリズム
        a_re, a_im = np.zeros_like(qx), np.zeros_like(qx)
        c_re, c_im = 0.0, 0.0
        for xy in self.pos_xy:
            delta = qx * xy[0] + qy * xy[1] # 位相
            re, im = np.cos(delta), np.sin(delta)
            y_re, y_im = re - c_re, im - c_im
            t_re, t_im = a_re + y_re, a_im + y_im
            c_re, c_im = (t_re - a_re) - y_re, (t_im - a_im) - y_im
            a_re, a_im = t_re, t_im
        self.a_re, self.a_im = a_re, a_im

        i_sim = (self.a_re**2 + self.a_im**2) * self.i_par
        sim_sum = (i_sim * self.mask).sum()
        self.i_sim = (i_sim * self.i_sum / sim_sum).astype(self.i_dtype)
        return

    def update_i(self, before_xy: np.ndarray, after_xy: np.ndarray):
        """直前に動かした粒子の情報を利用して散乱強度を更新する"""
        nx, ny, n_moved = self.qx.size, self.qy.size, before_xy.shape[0]
        qx, qy = np.meshgrid(self.qx, self.qy)
        # Kahanの加算アルゴリズム
        da_re, da_im = np.zeros_like(qx), np.zeros_like(qx) # 前後の振幅変化の総和
        c_re, c_im = 0.0, 0.0
        for before, after in zip(before_xy, after_xy):
            # 前後の位相
            delta_after = qx * after[0] + qy * after[1]
            delta_before = qx * before[0] + qy * before[1]
            # 前後の1粒子毎の振幅変化
            re = np.cos(delta_after) - np.cos(delta_before)
            im = np.sin(delta_after) - np.sin(delta_before)
            y_re, y_im = re - c_re, im - c_im
            t_re, t_im = da_re + y_re, da_im + y_im
            c_re, c_im = (t_re - da_re) - y_re, (t_im - da_im) - y_im
            da_re, da_im = t_re, t_im
        self.a_re += da_re
        self.a_im += da_im

        # 散乱強度(2次元配列)
        i_sim = (self.a_re**2 + self.a_im**2) * self.i_par
        sim_sum = (i_sim * self.mask).sum()
        self.i_sim = (i_sim * self.i_sum / sim_sum).astype(self.i_dtype)
        return

    def compute_residual(self):
        """実験値との残差を計算する"""
        # assert(self.i_sim.sum()/self.i_sum-1 < 1e-100)
        return np.sum((self.i_exp*self.mask - self.i_sim)**2)

    def __move(self, n_move: int, max_iter: int = -1) -> tuple[np.ndarray, np.ndarray]:
        """禁止域に入らないように粒子を動かす
        **pos_xyを更新しない**

        Parameters
        ----------
        n_move : int
            1ステップで動かす粒子の数
        max_iter : int, optional
            禁止域に入らないように粒子を動かす試行回数, by default 100

        Returns
        -------
        tuple[np.ndarray, np.ndarray]
            (移動前のxy座標の配列), (移動後のxy座標の配列)
        """
        if max_iter == -1:
            max_iter = self.MAX_RAND_TRIAL

        before_ab = self.pos_ab.copy()
        success = False
        idx = self.rs.choice(self.N, n_move, replace=False)
        for _ in range(max_iter):
            _success = False
            for i in idx:
                __success = False
                dirs = np.arange(6)
                self.rs.shuffle(dirs)
                for d in dirs:
                    new_ab = self.pos_ab[i] + self.STEP[d]
                    new_ab[0] = new_ab[0] % self.La
                    new_ab[1] = new_ab[1] % self.Lb
                    # 禁止域に入らないかチェック
                    m = (self.prohibited_a == new_ab[0]) * (self.prohibited_b == new_ab[1])
                    if m.sum() > 1: # 元の場所の近傍以外のprohibitedと一致したらアウト
                        continue
                    self.pos_ab[i] = new_ab
                    __success = True
                    break
                if not __success:
                    break
                _success = True

            # 失敗した場合は元に戻して再試行
            if not _success:
                self.pos_ab[idx] = before_ab[idx]
                idx = self.rs.choice(self.N, n_move, replace=False)
                continue

            success = True
            break

        if not success:
            raise RuntimeError("Failed to move particles")

        # idxには移動した粒子のインデックスが残っている
        before_xy = np.dot(before_ab[idx], self.conv)
        after_xy = np.dot(self.pos_ab[idx], self.conv)
        return before_xy, after_xy

    def move(self, n_move: int, max_iter: int = -1):
        """粒子を動かす

        Parameters
        ----------
        n_move : int
            1ステップで動かす粒子の数
        max_iter : int, optional
            禁止域に入らないように粒子を動かす試行回数, by default 100
        """
        self.__move(n_move, max_iter)
        self.pos_xy = np.dot(self.pos_ab, self.conv)
        return